api/routes/auth.js

```diff
@@ -0,0 +1,106 @@
+
+        const guest = await User.findOne({username})
+
+        //if user already exists, throw error
+        if (guest) throw new Error(`User: ${username} already exists!`)
+        
+        //store user in database
+        const saltRounds = 10
+        const hashed = await bcrypt.hash(password, saltRounds)
+        const user = await User.create({
+            username,
+            password: hashed
+        })
+        console.log(`User ${username} created!`)
+
+        //return success
+        const payload = { id: user._id } //setup payload
+        const options = { expiresIn: '1 day' } //add expiration
+        const token = jsonwebtoken.sign(payload, SECRET_KEY, options) //create token
+
+        res.status(status).json({status, token})
+
+    } catch (e) {
+        e.status = 400
+        next(e)
+    }
+})
+
+router.post('/login', async (req, res, next)=> {
+    const status = 201
+    try{
+        const { username, password } = req.body
+        //throw error if no username
+        if (!username) throw new Error(`Username required to login`)
+
+        const guest = await User.findOne({username})
+        const isValid = bcrypt.compare(password, guest.password)
+        //throw error if issue with username/password
+        if (!isValid) throw new Error(`Username and password do not match`)
+
+        const payload = { id: guest._id }
+        const options = { expiresIn: '1 day' }
+        const token = jsonwebtoken.sign(payload, SECRET_KEY, options)
+        res.status(status).json({status, token})
+    } catch (e) {
+        e.status = 401
+        next(e)
+    }
+})
+
+router.patch('/users/:id/permissions', async (req, res, next) => {
+    const status = 204
+    try {
+        //make sure request body is OK - 400
+        const { permissions } = req.body
+        const { id } = req.params
+        if ( permissions !== (true || false)) throw new Error (`There was a problem with your request body`)
+
+        //make sure headers contain auth - 401
+        const token = req.headers.authorization.split('Bearer ')[1]
+        if (!token) throw new Error(`There was a problem with your request`)
+        const payload = jsonwebtoken.verify(token, SECRET_KEY)
+
+        const user = await User.findOne({ _id: payload.id }).select('-__v -password')
+        //make sure request maker is an admin - 401
+        const { admin } = user 
+        if (admin !== true) throw new Error(`There was a problem with your request`)
+
+        //make sure user exists - 404
+        const updatedUser = await User.findOne({_id: id})
+        if (!updatedUser) throw new Error(`User ID: ${id} does not exist!`)
+
+        //need to verify this and make sure it is correct way to do..
+        updatedUser.admin = permissions
+        await updatedUser.save()
+
+        res.status(status).json({status})
+
+    } catch (e) {
+        if (e.message == 'There was a problem with your request body') {
```

Comment:
I don't know if this is the best way to capture the messages, although I see where you're going! Instead, maybe throw a "key" that you can interpret. For example:
```js
if (!updatedUser) throw new Error('user_does_not_exist')
// ...
if (e.message == 'user_does_not_exist') { /* ... */ }
```
This might be a bit easier in that if you want to change the message, you can do so more easily.